### ISSUE TITLE
"Unbound symbol" when using GADTs

### ISSUE THREAD

**Issue by pacastega** [ucsd-progsys/liquidhaskell#2349]
With the following code
```haskell
{-# LANGUAGE GADTs #-}
{-@ LIQUID "--reflection" @-}
module TestModule where

data Expr t where
  I :: Int -> Expr Int

  NIL :: Expr [t]
  CONS :: Int -> Expr t -> Expr [t] -> Expr [t]

{-@
data Expr t where
  I :: Int -> Expr Int

  NIL :: Expr {v:[t] | len v = 0}
  CONS :: l:Nat -> head:Expr t -> tail:Expr {v:[t] | len v = l} ->
          Expr {v:[t] | len v = l+1}
@-}
```
I am getting an error about `l` being unbound:
```
    Illegal type specification for `TestModule.$WCONS`
    TestModule.$WCONS :: forall t .
                         x1:GHC.Types.Int -> head##TestModule.CONS:(TestModule.Expr t) -> tail##TestModule.CONS:(TestModule.Expr {VV : [t] | len VV == l##TestModule.CONS}) -> {VV : (TestModule.Expr {VV : [t] | len VV == l##TestModule.CONS + 1}) | is$TestModule.CONS VV
                                                                                                                                                                                                                                                       && not (is$TestModule.I VV)
                                                                                                                                                                                                                                                       && not (is$TestModule.NIL VV)
                                                                                                                                                                                                                                                       && lqdc##$select##TestModule.CONS##1 VV == x1
                                                                                                                                                                                                                                                       && lqdc##$select##TestModule.CONS##2 VV == head##TestModule.CONS
                                                                                                                                                                                                                                                       && lqdc##$select##TestModule.CONS##3 VV == tail##TestModule.CONS}
    Sort Error in Refinement: {VV : [t##X0] | len VV == l##TestModule.CONS}
    Unbound symbol l##TestModule.CONS --- perhaps you meant: TestModule.CONS, is$TestModule.CONS ?
    Just constructor
  |
9 |   CONS :: Int -> Expr t -> Expr [t] -> Expr [t]
  |   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

**Comment by facundominguez** [comment]
The error is not helpful, but LH doesn't check the output of GADTs at the moment. See for instance [#2090](https://github.com/ucsd-progsys/liquidhaskell/issues/2090). Even if you could write such a specification, then LH wouldn't check the type of the result.

**Comment by nikivazou** [comment]
Actually, we have a short discussion with @goldfirere at ICFP about a related issue. 

The real problem here is that the GHC and liquid type for the worker types of the GADT worker constructors (e.g., `$WCONS`) do not match so the refinements are ignored. So, it would be good to understand the GHC rules for the generation of such types and update the generation of the refinement types.

**Comment by goldfirere** [comment]
Happy to provide the insider knowledge about GHC here. I *think* the key bit is that GHC translates definitions like

```
data T a where
  MkT :: T Int
```

to become

```
data T a where
  MkT :: forall a. (a ~# Int) => T a
```

The algorithm to create the new ("rejigged") type of the constructor (example below):
1. Zip all the datatype parameters (as declared in the type header, e.g. `data T a where`) with the indices in the return type of the constructor.
2. If the index in the return type of the constructor is a type variable that has not yet appeared in another index in the return type, it is considered a renaming; remember this.
3. Quantify all the universal variables with foralls. These are the variables from the header, in the order given there, but any renamed variables (from step 2) have their new names.
4. Quantify any variables free in the type of the constructor that aren't just renamed universals. These are the existentials.
5. Include unlifted equalities (built with `~#`) mapping each unrenamed universal variable with its corresponding index type from the constructor return type.
6. The argument types to the constructor don't change.
7. The result type is the type constructor followed by all the (possibly renamed) universals.

Bigger example:

```
data T a b c where
  MkT :: T Int q q
```

becomes

```
data T a b c where
  MkT :: forall a q c. (a ~# Int) => (c ~# q) => T a q c
```

All of this happens for the *worker* type. The *wrapper* type is, I believe, the type as written by the user. My guess is that LH uses the worker type (or you probably wouldn't be asking this question).

Does this help?

**Comment by facundominguez** [comment]
Fixed by #2370
